libbindgen/src/ir/item.rs

```diff
@@ -525,6 +526,107 @@ impl Item {
         self.as_type().map_or(false, |ty| ty.is_type_ref())
     }
 
+    /// Get the target item id for name generation.
+    fn name_target(&self, ctx: &BindgenContext, for_name_checking: bool) -> ItemId {
+        let mut item = self;
+        loop {
+            match *item.kind() {
+                ItemKind::Type(ref ty) => {
+                    match *ty.kind() {
+                        // If we're a template specialization, our name is our
+                        // parent's name.
+                        TypeKind::Comp(ref ci) if ci.is_template_specialization() => {
+                            item = ctx.resolve_item(ci.specialized_template().unwrap());
+                        },
+                        // Same as above.
+                        TypeKind::ResolvedTypeRef(inner) |
+                        TypeKind::TemplateRef(inner, _) => {
+                            item = ctx.resolve_item(inner);
+                        }
+                        // Template aliases use their inner alias type's name if we
+                        // are checking names for whitelisting/replacement/etc.
+                        TypeKind::TemplateAlias(inner, _) if for_name_checking => {
+                            item = ctx.resolve_item(inner);
+                        }
+                        _ => return item.id(),
+                    }
+                },
+                _ => return item.id(),
+            }
+        }
+    }
+
+    /// Get this function item's name, or `None` if this item is not a function.
+    fn func_name(&self) -> Option<&str> {
+        if let ItemKind::Function(ref func) = *self.kind() {
+            Some(func.name())
+        } else {
+            None
+        }
+    }
+
+    /// Get the overload index for this method. If this is not a method, return
+    /// `None`.
+    fn method_overload_index(&self, ctx: &BindgenContext) -> Option<usize> {
```

Comment:
Ahh, I understand why this can't be completely moved to codegen, we generate calls for it... Well looks fine :)